### PR TITLE
Add binary parser

### PR DESCRIPTION
**Pull request by tthug**
Adding a binary parser module to support deserializing a transaction from its binary form.
This parser supports all fields that are being serialized. 
_NOTE_: Parsing of amounts in other formats than drops is currently not supported but should be easily extandable onto these changes.

### CODE REVIEW

**Comment by gmosx** [comment]
Thank you for the PR. I will review tomorrow!

File: xrpl_types/src/transaction.rs [hunk 1]
```diff
@@ -27,6 +27,12 @@ pub enum TransactionType {
     DepositPreauth = 19,
     TrustSet = 20,
     AccountDelete = 21,
+    SetHook = 22,
```
Comment (gmosx) [inline xrpl_types/src/transaction.rs:30]: Where did you find these values (link)?

File: xrpl_binary_codec/src/parser.rs [hunk 2]
```diff
@@ -0,0 +1,616 @@
+use xrpl_types::{Amount, Memo, Transaction, TransactionType};
+
+use crate::{
+    error::BinaryCodecError,
+    serializer::{HASH_PREFIX_TRANSACTION, HASH_PREFIX_UNSIGNED_TRANSACTION_SINGLE},
+};
+
+#[derive(Default)]
+pub struct Parser {
```
Comment (gmosx) [inline xrpl_binary_codec/src/parser.rs:9]: Nitpick: Could be named deserializer to be symmetric with serializer? Not a blocker, but we might consider renaming this in a _followup_ PR.

File: xrpl_binary_codec/src/parser.rs [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+        if self.bytes.is_empty() {
+            Err(BinaryCodecError::ParseError(
+                "Unable to read byte from empty buffer".to_string(),
+            ))
+        } else {
+            Ok(self.bytes.remove(0))
+        }
+    }
+
+    fn read_u_int_16(&mut self) -> Result<u16, BinaryCodecError> {
```
Comment (gmosx) [inline xrpl_binary_codec/src/parser.rs:418]: Nitpick: u_int -> uint? What do you think? (not a blocker).

File: xrpl_binary_codec/src/parser.rs [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+        // Pop the bytes in reverse order so we get the bytes in correct order
+        if self.bytes.len() >= 4 {
+            let p1 = self.bytes.remove(0);
+            let p2 = self.bytes.remove(0);
+            let p3 = self.bytes.remove(0);
+            let p4 = self.bytes.remove(0);
+            Ok(((p1 as u32) << 24) | ((p2 as u32) << 16) | ((p3 as u32) << 8) | p4 as u32)
+        } else {
+            Err(BinaryCodecError::ParseError(
+                "Unable to ready 32 bit integer with less then 4 bytes left of the buffer"
```
Comment (gmosx) [inline xrpl_binary_codec/src/parser.rs:442]: ready -> read?

**Comment by gmosx** [comment]
I think I will merge it as is, and we can refine in follow-up PRs, if needed.

File: xrpl_types/src/transaction.rs (hunk 1 above)
Comment (tthug) [reply xrpl_types/src/transaction.rs:30]: I found them here: https://github.com/XRPLF/xrpl.js/blob/main/packages/ripple-binary-codec/src/enums/definitions.json#L2303 

and confirmed here: https://github.com/XRPLF/rippled/blob/e32bc674aa2a035ea0f05fe43d2f301b203f1827/src/ripple/protocol/TxFormats.h#L56
will ad these sources as an inline comment in a follow up PR

File: xrpl_binary_codec/src/parser.rs (hunk 2 above)
Comment (tthug) [reply xrpl_binary_codec/src/parser.rs:9]: True, I was going to do that but then saw that there seems to be a community convention of calling it a Serializer and a Parser. At least in [xrpls](https://github.com/XRPLF/xrpl.js/tree/main/packages/ripple-binary-codec/src/serdes) and [xrpl4j](https://github.com/XRPLF/xrpl4j/blob/main/xrpl4j-binary-codec/src/main/java/org/xrpl/xrpl4j/codec/binary/serdes/BinaryParser.java) strange but I thought it best to do as the others. Still, open for renaming this to deserializer it is more intuative for what this is.

File: xrpl_binary_codec/src/parser.rs (hunk 3 above)
Comment (tthug) [reply xrpl_binary_codec/src/parser.rs:418]: agreed, `uint` is cleaner

File: xrpl_types/src/transaction.rs (hunk 1 above)
Comment (gmosx) [reply xrpl_types/src/transaction.rs:30]: Thanks.

File: xrpl_binary_codec/src/parser.rs (hunk 2 above)
Comment (gmosx) [reply xrpl_binary_codec/src/parser.rs:9]: OK, let's think some more about it. I am wondering what is the idiomatic Rust names.

(We could even name the coder/decoder, from codec)

**Pull request merged**